### DISCUSSION TITLE
Report | Week 05

### DISCUSSION THREAD

**Discussion by tech-n1c** [SE-TINF22B6/time2shine#27] [Show and tell]
## Overview
This week we worked on the setup of our CI/CD Environment. We use GitHub Actions to build and deploy our Spring Boot Backend as well as the frontend. On every change on our main branch, the pipelines are triggered. After the build (only for the backend) ist finished, everything gets deployed in our Digital Ocean instance (thanks to @Shiny1708 for the recommendation 😊).

---

### GitHub Actions Backend Pipeline
When triggered, the Gradle project is being built in the Actions pipeline. In a second step, we execute the tests explained further down in the document. Then, we build a Docker Image from the resulting .jar archive. This image is then pushed to a Dockerhub repository. If you're interested, you can find the config here: [click](https://github.com/SE-TINF22B6/time2shine/blob/main/.github/workflows/backend-pipeline.yml)

![image](https://github.com/SE-TINF22B6/time2shine/assets/59262249/b84547a5-3d05-43fe-91a4-ec969c8259dc)

### GitHub Actions Frontend Pipeline
The frontend pipeline is much simpler, as the frontend only consists of HTML, CSS and Javascript files that don't have to be compiled or something like that. Therefore, we can directly build the Docker Image and push it to Dockerhub the same way as for the backend. If you're interested, you can find the config here: [click](https://github.com/SE-TINF22B6/time2shine/blob/main/.github/workflows/frontend-pipeline.yml)

![image](https://github.com/SE-TINF22B6/time2shine/assets/59262249/1812091d-715a-468d-b1d3-848895e9d568)

### DockerHub
This is where we maintain our Docker Images for the backend and frontend. This way, we can easily push new versions from GitHub and also pull the latest one from our Digital Ocean instance.

### Deployment at Digital Ocean
Digital Ocean provides Web Hosting services, it is where we have an instance of Debian 12 with Docker running to deploy the different components of our application. As previously mentioned, @Shiny1708 recommended it to us, because students can get a 200 $ credit to spend on their different services. This is very cool, because the bwCloud for instance only provides very slow instances for free. But make sure not to exceed the 200 $, because the prices are quite high compared to other providers (others don't give the free credit however)!

---

## JUnit Backend Testing
We have implemented simple JUnit tests for account management to check basic functionality, including account creation, deletion and data retrieval. In the near future, we plan to add more tests to further ensure the reliability and security of the system.
The testing process is also integrated in the backend pipeline (as previously described)

---

![onemorething](https://github.com/SE-TINF22B6/time2shine/assets/59262249/cdbd6d3f-9b09-4b72-b4e1-376c08d79c25)

GitHub has a built-in feature to calculate the amount of different programming languages (and also HTML) used in your repository, which is called Github-Linguist. You can find it [here](https://github.com/github-linguist/linguist) if you encounter any problems with it. We needed to add a [.gitattributes file](https://github.com/SE-TINF22B6/time2shine/blob/main/.gitattributes) to exclude our Pixi.JS library from the calculation. Because the library is very big, it made the code calculation completely wrong, but this can be fixed using the .gitattributes file. So if you have similar problems, check out the Linguist repo or feel free to ask us.
![image](https://github.com/SE-TINF22B6/time2shine/assets/59262249/3e5a0285-ce52-4908-a43f-17380da33ff0)

![image](https://github.com/SE-TINF22B6/time2shine/assets/59262249/83115a7f-88e2-4cc7-ac85-a2e7f85b2dee)

**Comment by BenjaminB109** [comment]
Dear time2shine Team,
Once again a good article. Especially the part about Digital Ocean is very interesting. We will discuss in the team whether we will also use this.

**Comment by Nahlam4** [comment]
Dear Team time2shine,
Thank you for the Overview of your CI/CD Environment and your Pipelines. I especially like that you explained the Systems and tools that
you´re using. Just one small adjustment: the Link to your Backend Pipeline isn´t up to date.

**Reply by tech-n1c** [reply]
Thanks for the hint, the link should be fixed now.

**Comment by MaarcS** [comment]
Very informative blog post. I really like how you automated the deployment to digital ocean, and I am sure that this will make your workflow a lot faster. You said your frontend pipeline is very simple, maybe you could also include some frontend tests (if possible) for additional quality assurance.
Also _**thank you**_ for the tip for GitHub-Linguist, it always annoyed me when it included all my javascript libraries. I'm glad that you found a fix for that.
And as Josha already said, one of your links is broken :)

**Reply by tech-n1c** [reply]
Thank you for your feedback! You're absolutely right about testing in the frontend, as a lot of game logic will happen there. But right now, the frontend doesn't really consist of anything that could be tested. But we'll make sure to do this in the future!

**Comment by derrrobin** [comment]
Hey there, Your post is informative and provides a good overview of your CI/CD environment setup using GitHub Actions, DockerHub, and Digital Ocean. Here are some positive aspects and suggestions for improvement:

Positive Points:

 The post is well-structured and concise, making it easy to follow the different stages of your CI/CD process.
You've included links to your GitHub Actions configuration, which is helpful for readers who want to dive deeper into your setup.

A feedback point form me would be to consider adding a concluding section that summarizes the key takeaways or encourages readers to ask questions or seek further information.

Your CookPit Team

**Comment by maxschwinghammer** [comment]
Dear Team time2shine,
Great overview! The utilization of GitHub Actions for CI/CD setup is well-explained. The use of Digital Ocean for hosting seems like a practical choice, especially with the instance running Docker. I think we will consider using this in our team too. Do you have any experience of how quickly the $200 student credit is used up or how long it is likely to last?

**Comment by AndresMasis** [comment]
Hi.

First of all I find your workflow very interesting. Because you are using Github Actions as us, but you are using Spring, which is a different technology than ours. So this is a very valuable comparison of how different technologies connect with the same CI/CD tool.
Also I found very useful your explanation of Github-Linguist. The explanation of how to manipulate the .gitattributes was something that I di not know and know I can apply for this project.

Cheers from the CookPit team.
